server.go

```diff
@@ -191,6 +191,43 @@ func New(opts *Options) *Server {
 func (o *Options) ensureDefaults() {
 	if o.Capabilities == nil {
 		o.Capabilities = DefaultServerCapabilities
+	} else {
+		if o.Capabilities.MaximumSessionExpiryInterval == 0 {
+			o.Capabilities.MaximumSessionExpiryInterval = DefaultServerCapabilities.MaximumSessionExpiryInterval
+		}
+		if o.Capabilities.MaximumMessageExpiryInterval == 0 {
+			o.Capabilities.MaximumMessageExpiryInterval = DefaultServerCapabilities.MaximumMessageExpiryInterval
```

Comment:
So this is actually a tricky one. Because none of these are pointers, both the user input value or 0 and the zero value are indistinguishable. I will have to think on this one

Reply:
Nice to see you finally running into all the same problems I ran into 😆 

I ended up using pointers for packet properties to get around this issue.

Reply:
@dgduncan It's complicated. Code wise it's the correct solution. Usability wise, it's not great because it requires the user to set a pointer to an int, which is a pain (AWS SDK does this and I never got on with it). We could add methods to set the values for the config struct, but that's not very idiomatic go. 

Another solution is to add a `bool EnforceMaximumMessageExpiryInterval` property and continue using non-pointer ints.

Reply:
Upon looking further the problem is on many of the `capabilities`. Pretty much any `byte` type field will incorrect if set to 0.

Reply:
Ok so I think I may have an idea. I am starting to lean towards leaving the defaults as is and putting the burden of doing these checks onto the `configs` themselves. For example, with the `yaml` configurations it is very easy to detect when a value is not set. So the "smart" defaults can be done there. 

I simply feel that the previous changes with the "smart" defaults as suggested here simply make too much noise for little benefit.

Reply:
For anyone interested, please take a look at the changes I made in another [PR](https://github.com/mochi-mqtt/server/blob/7359f24b1dc0e023d08ee7c41d9de9cff051090e/configs/file/options.go)

This creates a duplicate `Options` that lives in the `config` package that has everything marked as a pointer. That way, you can check each value for `nil` and if it is NOT `nil` (i.e the user has set it in the config) it will overwrite the default value generated from the newly exported `EnsureDefaults`. I will change the names of some of these things as they do not make sense at the moment. 

Advantages
1. Minimal changes to core broker code
2. Default checking is done by the `config` package itself thusly allowing for more robust checks
3. Although not done yet, allows for the removal of the `yaml` tags from the `mqtt.Options`.

Disadvantages
1. Creates a duplicate `Options` that can become out of sync
2. Not great in terms of code duplication

Reply:
@mochi-co I agree with you, and lean towards adding a field like 'MaximumMessageExpiryDisable' to determine whether to disable the forced message expiration, rather than relying on whether the value of 'MaximumMessageExpiryInterval' == 0. This will make the code clearer.

Reply:
If we need to enable 'MaximumMessageExpiry' by default, I think naming the variable 'MaximumMessageExpiryDisable' would be better because the default value for a boolean type is 'false'.

Reply:
@werbenhu I can't speak on the conversation in PR #315 as I have not been following it directly; however, in regards to this specific PR I believe I may have come up with a better way to deal with default configs as show in PR #309 
I personally do not mind the `MaximumMessageExpiry == 0` as that is how it is commonly done in other brokers and when connecting in many mqtt clients but I do not have a strong opinion

Reply:
> I believe I may have come up with a better way to deal with default configs as show in PR https://github.com/mochi-mqtt/server/pull/309

I'm still not 100% convinced about the current structure of that PR - I may make some code suggestions later if get time later as I have some opinions on how we can extend your concept, if that's okay.

> I personally do not mind the MaximumMessageExpiry == 0 as that is how it is commonly done in other brokers and when connecting in many mqtt clients but I do not have a strong opinion

As far as I can see, in this change it's impossible to set the `MaximumMessageExpiry = 0` when passing the parameter in the server options:

```go
server := mqtt.New(&mqtt.Options{
	Capabilities: &mqtt.Capabilities{
		MaximumMessageExpiryInterval: 0,
	},
})
```

As it will immediately be replaced by `ensureDefaults` with `DefaultServerCapabilities.MaximumMessageExpiryInterval` (86400)

You can set it after the fact, but I think this might be confusing and counter-intuitive for people trying to use it.

If we are going to move forward with file based configurations, this will (in my mind) involve changing the way configurations are passed to the server and potentially a major SemVer increase. My feeling is that we should implement this change into that same project, as any changes we make in this task will at the very least be a minor SemVer increase and may require additional changes following #309. 

My recommendation would be to close this PR and incorporate it (and code if necessary) into #309 as a new requirement. 

Also wondering if we can make a branch on mochi-mqtt/server that we can all push to within the different projects 🤔 I'll investigate

Reply:
@mochi-co Sounds good I will close this PR. I am having some trouble wrapping my head around the best way to incorporate what is essentially a wrapper around the `Options` configuration without it looking pretty clunky.

I guess in my eyes I like the idea of any "configs" wrappers, i.e file based configuration to basically live it it's own little world as it is not `core` to the broker functionality. In doing this, I am personally ok with it occasionally becoming out of step with changes in the broker. I would rather it be this way then have the broker configuration change to the needs of outside configuration. To me it is like when working on a vendor product, sometime the CLI becomes out of date compared to the API or SDK. It is not unexpected.

Reply:
@dgduncan I think the way the core server implements configuration is wrong to start with. If it were implemented correctly in the first place we would be able to translate file configs directly. We can get ourselves tied up in knots trying to build solutions for tricky problems when in fact it's the initial implementation is what needs fixing 🙂 I'll see if I can get a proposal  PR together this week